File: archivist/src/main/java/com/zorroa/archivist/web/api/FolderController.java
package com.zorroa.archivist.web.api;

import com.zorroa.archivist.HttpUtils;
import com.zorroa.archivist.domain.Folder;
import com.zorroa.archivist.domain.FolderSpec;
import com.zorroa.archivist.service.AssetService;
import com.zorroa.archivist.service.FolderService;
import com.zorroa.archivist.service.MessagingService;
import com.zorroa.archivist.service.SearchService;
import com.zorroa.sdk.search.AssetSearch;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.servlet.HandlerMapping;

import javax.servlet.http.HttpServletRequest;
import java.util.List;

@RestController
public class FolderController {

    @Autowired
    FolderService folderService;

    @Autowired
    AssetService assetService;

    @Autowired
    SearchService searchService;

    @Autowired
    MessagingService messagingService;

    @RequestMapping(value="/api/v1/folders", method=RequestMethod.POST)
    public Folder create(@RequestBody FolderSpec spec) {
        return folderService.create(spec, false);
    }

    @RequestMapping(value="/api/v1/folders", method=RequestMethod.GET)
    public List<Folder> getAll() {
        return folderService.getAll();
    }

    @RequestMapping(value="/api/v1/folders/{id:\\d+}", method=RequestMethod.GET)
    public Folder get(@PathVariable int id) {
        return folderService.get(id);
    }

    @RequestMapping(value="/api/v1/folders/{id}/_assetCount", method=RequestMethod.GET)
    public Object countAssets(@PathVariable int id) {
        return HttpUtils.count(searchService.count(folderService.get(id)));
    }

    public static class FolderAssetCountsRequest {
        public List<Integer> ids;
        public AssetSearch search;
    }

    @RequestMapping(value="/api/v1/folders/_assetCounts", method=RequestMethod.POST)
    public Object countAssets(@RequestBody FolderAssetCountsRequest req) {
        return HttpUtils.counts(searchService.count(req.ids, req.search));
    }

    @RequestMapping(value="/api/v1/folders/_path/**", method=RequestMethod.GET)
    public Folder get(HttpServletRequest request) {
        String path = (String) request.getAttribute(
                HandlerMapping.PATH_WITHIN_HANDLER_MAPPING_ATTRIBUTE);
        path = path.substring(path.indexOf("/_path/") + 6).replace("//", "/");
        Folder f =  folderService.get(path);
        return f;
    }

    @Deprecated
    @RequestMapping(value="/api/v1/folders/_exists/**", method=RequestMethod.GET)
    public Object exists(HttpServletRequest request) {
        String path = (String) request.getAttribute(
                HandlerMapping.PATH_WITHIN_HANDLER_MAPPING_ATTRIBUTE);
        path = path.replace("/api/v1/folders/_exists", "");
        return HttpUtils.status("folders", path, "exists", folderService.exists(path));
    }

    @RequestMapping(value="/api/v1/folders/{id}", method=RequestMethod.PUT)
    public Folder update(@RequestBody Folder folder, @PathVariable int id) {
        folderService.update(id, folder);
        return folderService.get(folder.getId());
    }

    @RequestMapping(value="/api/v1/folders/{id}", method=RequestMethod.DELETE)
    public Object delete(@PathVariable int id) {
        Folder folder = folderService.get(id);
        return HttpUtils.deleted("folders", id, folderService.trash(folder).getCount() > 0);
    }

    @RequestMapping(value="/api/v1/folders/{id}/folders", method=RequestMethod.GET)
    public List<Folder> getChildFolders(@PathVariable int id) {
        Folder folder = folderService.get(id);
        return folderService.getChildren(folder);
    }

    @Deprecated
    @RequestMapping(value="/api/v1/folders/{id}/_children", method=RequestMethod.GET)
    public List<Folder> getChildren(@PathVariable int id) {
        Folder folder = folderService.get(id);
        return folderService.getChildren(folder);
    }

    @RequestMapping(value="/api/v1/folders/{id}/folders/{name}", method=RequestMethod.GET)
    public Folder getChild(@PathVariable int id, @PathVariable String name) {
        return folderService.get(id, name);
    }

    /**
     * Remove the given list of asset Ids from a folder.
     *
     * @param assetIds
     * @param id
     * @throws Exception
     */
    @RequestMapping(value="/api/v1/folders/{id}/assets", method=RequestMethod.DELETE)
    public Object removeAssets(
            @RequestBody List<String> assetIds,
            @PathVariable Integer id) throws Exception {
        Folder folder = folderService.get(id);
        return folderService.removeAssets(folder, assetIds);
    }

    /**
     * Add a given list of asset Ids from a folder.
     *
     * @param assetIds
     * @param id
     * @throws Exception
     */
    @RequestMapping(value="/api/v1/folders/{id}/assets", method=RequestMethod.POST)
    public Object addAssets(
            @RequestBody List<String> assetIds,
            @PathVariable Integer id) throws Exception {
        Folder folder = folderService.get(id);
        return folderService.addAssets(folder, assetIds);
    }
}

File: archivist/src/main/kotlin/com/zorroa/archivist/web/api/FolderController.kt
package com.zorroa.archivist.web.api

import com.zorroa.archivist.HttpUtils
import com.zorroa.archivist.domain.Folder
import com.zorroa.archivist.domain.FolderSpec
import com.zorroa.archivist.domain.SetPermissions
import com.zorroa.archivist.service.FolderService
import com.zorroa.archivist.service.SearchService
import com.zorroa.sdk.search.AssetSearch
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation.*
import org.springframework.web.servlet.HandlerMapping
import javax.servlet.http.HttpServletRequest

@RestController
class FolderController @Autowired constructor(
        private val folderService: FolderService,
        private val searchService: SearchService
) {

    val all: List<Folder>
        @GetMapping(value = "/api/v1/folders")
        get() = folderService.getAll()

    @PostMapping(value = "/api/v1/folders")
    fun create(@RequestBody spec: FolderSpec): Folder {
        return folderService.create(spec, false)
    }

    @GetMapping(value = "/api/v1/folders/{id:\\d+}")
    operator fun get(@PathVariable id: Int): Folder {
        return folderService.get(id)
    }

    @GetMapping(value = "/api/v1/folders/{id}/_assetCount")
    fun countAssets(@PathVariable id: Int): Any {
        return HttpUtils.count(searchService.count(folderService.get(id)))
    }

    class FolderAssetCountsRequest {
        var ids: List<Int>? = null
        var search: AssetSearch? = null
    }

    @PostMapping(value = "/api/v1/folders/_assetCounts")
    fun countAssets(@RequestBody req: FolderAssetCountsRequest): Any {
        return HttpUtils.counts(searchService.count(req.ids!!, req.search))
    }

    @GetMapping(value = "/api/v1/folders/_path/**")
    fun get(request: HttpServletRequest): Folder? {
        var path = request.getAttribute(
                HandlerMapping.PATH_WITHIN_HANDLER_MAPPING_ATTRIBUTE) as String
        path = path.substring(path.indexOf("/_path/") + 6).replace("//", "/")
        return folderService.get(path)
    }

    @Deprecated("")
    @GetMapping(value = "/api/v1/folders/_exists/**")
    fun exists(request: HttpServletRequest): Any {
        var path = request.getAttribute(
                HandlerMapping.PATH_WITHIN_HANDLER_MAPPING_ATTRIBUTE) as String
        path = path.replace("/api/v1/folders/_exists", "")
        return HttpUtils.status("folders", path, "exists", folderService.exists(path))
    }

    @PutMapping(value = "/api/v1/folders/{id}")
    fun update(@RequestBody folder: Folder, @PathVariable id: Int): Folder {
        folderService.update(id, folder)
        return folderService.get(folder.id)
    }

    @DeleteMapping(value = "/api/v1/folders/{id}")
    fun delete(@PathVariable id: Int): Any {
        val folder = folderService.get(id)
        return HttpUtils.deleted("folders", id, folderService.trash(folder).count > 0)
    }

    @GetMapping(value = "/api/v1/folders/{id}/folders")
    fun getChildFolders(@PathVariable id: Int): List<Folder> {
        val folder = folderService.get(id)
        return folderService.getChildren(folder)
    }

    @Deprecated("")
    @GetMapping(value = "/api/v1/folders/{id}/_children")
    fun getChildren(@PathVariable id: Int): List<Folder> {
        val folder = folderService.get(id)
        return folderService.getChildren(folder)
    }

    @GetMapping(value = "/api/v1/folders/{id}/folders/{name}")
    fun getChild(@PathVariable id: Int, @PathVariable name: String): Folder {
        return folderService.get(id, name)
    }

    @PutMapping(value = "/api/v1/folders/{id}/_permissions")
    @Throws(Exception::class)
    fun setPermissions(@PathVariable id: Int, @RequestBody req: SetPermissions): Any {
        val folder = folderService.get(id)
        if (req.replace) {
            folderService.setAcl(folder, req.acl!!, false, false)
        } else {
            folderService.updateAcl(folder, req.acl!!)
        }
        return folderService.get(folder.id)
    }

    /**
     * Remove the given list of asset Ids from a folder.
     *
     * @param assetIds
     * @param id
     * @throws Exception
     */
    @DeleteMapping(value = "/api/v1/folders/{id}/assets")
    @Throws(Exception::class)
    fun removeAssets(
            @RequestBody assetIds: List<String>,
            @PathVariable id: Int?): Any {
        val folder = folderService.get(id!!)
        return folderService.removeAssets(folder, assetIds)
    }

    /**
     * Add a given list of asset Ids from a folder.
     *
     * @param assetIds
     * @param id
     * @throws Exception
     */
    @PostMapping(value = "/api/v1/folders/{id}/assets")
    @Throws(Exception::class)
    fun addAssets(
            @RequestBody assetIds: List<String>,
            @PathVariable id: Int?): Any {
        val folder = folderService.get(id!!)
        return folderService.addAssets(folder, assetIds)
    }
}
